eessi/testsuite/eessi_mixin.py

```diff
@@ -0,0 +1,118 @@
+from reframe.core.pipeline import RegressionMixin
+from reframe.core.exceptions import ReframeSyntaxError
+
+from eessi.testsuite import hooks
+from eessi.testsuite.constants import DEVICE_TYPES, CPU, GPU, SCALES, COMPUTE_UNIT
+
+# Hooks from the Mixin class seem to be executed _before_ those of the child class
+# Thus, if the Mixin class needs self.X to be defined in after setup, the child class would have to define it before setup
+# That's a disadvantage and might not always be possible - let's see how far we get
+# It also seems that, like normal inheritance, functions with the same in the child and parent class will mean the child class
+# will overwrite that of the parent class. That is a plus, as we can use the EESSI_Mixin class as a basis, but still overwrite
+# specific functions in case specific tests would require this
+# TODO: for this reason, we probably want to put _each_ hooks.something invocation into a seperate function, so that each 
+# individual one can be overwritten
+
+# Note that I don't think we can do anything about the things set in the class body, such as the parameter's.
+# Maybe we can move those to an __init__ step of the Mixin, even though that is not typically how ReFrame does it anymore?
+# That way, the child class could define it as class variables, and the parent can use it in its __init__ method?
+class EESSI_Mixin(RegressionMixin):
+    """
+    All EESSI tests should derive from this mixin class unless they have a very good reason not to.
+    To run correctly, tests inheriting from this class need to define variables and parameters that are used here.
+    That definition needs to be done 'on time', i.e. early enough in the execution of the ReFrame pipeline.
+    Here, we list which class attributes need to be defined, and by (the end of) what phase:
+
+    - Init phase: device_type, scale, module_name
+    - Setup phase: compute_unit, required_mem_per_node
+    """
+
```

Comment:
we can define here the parameters that are required as "abstract parameters", e.g.
```
 module_name = parameter()
```
which ensures that the parameter is always set in the test.

Reply:
Yes, but the problem is that if the child class doesn't overwrite that, the error is _extremely_ unclear:

```
WARNING: skipping test 'EESSI_LAMMPS_lj': test has one or more undefined parameters
WARNING: skipping test 'EESSI_LAMMPS_rhodo': test has one or more undefined parameters
```
It might get better if this issue https://github.com/reframe-hpc/reframe/issues/3254 gets resolved. In that case, this approach is indeed much cleaner. The issue is on the backlog for the ReFrame 4.7 milestone, and I gave them a piece of code that _should_ do the job. So there is a fair chance we'll get it. In which case your approach is indeed cleaner.

The one downside (and this is a major one, since it is not trivial) is that we then cannot state _when_ we need that parameter. Some of these, the user might only be able to set after the `setup` phase, because they want to use information from the `current_partition`. They need to then know if that is 'early enough' for what the Mixin class requires (i.e. when the hooks called in the Mixin class require that variable to be set).

Reply:
stating which parameter is undefined would indeed be a welcome improvement.

however, `parameter`s should always be defined in the class body, there is no doubt when to set it.
see also the discussion in slack: https://reframetalk.slack.com/archives/CF8ENLUDB/p1725607966946879

> For now, I'd like to go for the verbose option of raising the error ourselves, so that we control what it says.

i'm fine with that, i also think an error is better than a warning here. so this should be done in the class body of the mixin class, right?

for normal or user-defined variables, we'll probably have to roll our own in any case indeed.